js/bitopro.js

```diff
@@ -0,0 +1,1349 @@
+        //             }
+        //         ]
+        //     }
+        //
+        const result = {};
+        for (let i = 0; i < currencies.length; i++) {
+            const currency = currencies[i];
+            const currencyId = this.safeString (currency, 'currency');
+            const code = this.safeCurrencyCode (currencyId);
+            const deposit = this.safeValue (currency, 'deposit');
+            const withdraw = this.safeValue (currency, 'withdraw');
+            const fee = this.safeNumber (currency, 'withdrawFee');
+            const withdrawMin = this.safeNumber (currency, 'minWithdraw');
+            const withdrawMax = this.safeNumber (currency, 'maxWithdraw');
+            const limits = {
+                'withdraw': {
+                    'min': withdrawMin,
+                    'max': withdrawMax,
+                },
+                'amount': {
+                    'min': undefined,
+                    'max': undefined,
+                },
+            };
+            result[code] = {
+                'id': currencyId,
+                'code': code,
+                'info': currency,
+                'type': undefined,
+                'name': undefined,
+                'active': deposit && withdraw,
+                'deposit': deposit,
+                'withdraw': withdraw,
+                'fee': fee,
+                'precision': undefined,
+                'limits': limits,
+            };
+        }
+        return result;
+    }
+
+    async fetchMarkets (params = {}) {
+        const response = await this.publicGetProvisioningTradingPairs ();
+        const markets = this.safeValue (response, 'data', []);
+        //
+        //     {
+        //         "data":[
+        //             {
+        //                 "pair":"shib_twd",
+        //                 "base":"shib",
+        //                 "quote":"twd",
+        //                 "basePrecision":"8",
+        //                 "quotePrecision":"6",
+        //                 "minLimitBaseAmount":"100000",
+        //                 "maxLimitBaseAmount":"5500000000",
+        //                 "minMarketBuyQuoteAmount":"1000",
+        //                 "orderOpenLimit":"200",
+        //                 "maintain":false,
+        //                 "orderBookQuotePrecision":"6",
+        //                 "orderBookQuoteScaleLevel":"5"
+        //             }
+        //         ]
+        //     }
+        //
+        const result = [];
+        for (let i = 0; i < markets.length; i++) {
+            const market = markets[i];
+            const active = !this.safeValue (market, 'maintain');
+            const pair = this.safeString (market, 'pair');
+            const baseId = this.safeString (market, 'base');
+            const quoteId = this.safeString (market, 'quote');
+            const id = pair;
+            let base = this.safeCurrencyCode (baseId);
+            base = base.toUpperCase ();
```

Comment:
we don't need this line since safeCurrencyCode will already make the codes uppercase